beaker/src/stamp_manager.rs

```diff
@@ -0,0 +1,367 @@
+        "alpha_matting_foreground_threshold:{}",
+        config.alpha_matting_foreground_threshold
+    ));
+    hasher.update(format!(
+        "alpha_matting_background_threshold:{}",
+        config.alpha_matting_background_threshold
+    ));
+    hasher.update(format!(
+        "alpha_matting_erode_size:{}",
+        config.alpha_matting_erode_size
+    ));
+    hasher.update(format!("save_mask:{}", config.save_mask));
+
+    if let Some(bg_color) = config.background_color {
+        hasher.update(format!("background_color:{bg_color:?}"));
+    }
+
+    // Include custom model parameters if specified
+    if let Some(ref model_path) = config.model_path {
+        hasher.update(format!("model_path:{model_path}"));
+    }
+    if let Some(ref model_url) = config.model_url {
+        hasher.update(format!("model_url:{model_url}"));
+    }
+    if let Some(ref model_checksum) = config.model_checksum {
+        hasher.update(format!("model_checksum:{model_checksum}"));
+    }
+
+    // Include output directory if it affects output file names/paths
+    if let Some(ref output_dir) = config.base.output_dir {
+        hasher.update(format!("output_dir:{output_dir}"));
+    }
+
+    format!("{:x}", hasher.finalize())[..16].to_string()
+}
+
+/// Generate tool version hash
+pub fn generate_tool_hash() -> String {
+    let mut hasher = Sha256::new();
+    hasher.update(format!("beaker:{}", env!("CARGO_PKG_VERSION")));
+    format!("{:x}", hasher.finalize())[..16].to_string()
+}
+
+/// Generate model hash from file path
+pub fn generate_model_hash_from_path(model_path: &Path) -> Result<String> {
+    if model_path.exists() {
+        calculate_md5(model_path).map(|hash| hash[..16].to_string())
+    } else {
+        Err(anyhow!("Model file not found: {}", model_path.display()))
+    }
+}
+
+/// Get the beaker stamps cache directory
+pub fn get_stamps_cache_dir() -> Result<PathBuf> {
+    let cache_dir = get_cache_base_dir()?;
+    Ok(cache_dir.join("beaker").join("stamps"))
+}
+
+/// Create stamp file if content has changed
+pub fn create_or_update_stamp(stamp_type: &str, hash: &str, content: &str) -> Result<PathBuf> {
+    let stamps_dir = get_stamps_cache_dir()?;
+    fs::create_dir_all(&stamps_dir)?;
+
+    let stamp_filename = format!("{stamp_type}-{hash}.stamp");
+    let stamp_path = stamps_dir.join(stamp_filename);
+
+    // Check if stamp already exists with same content
+    if stamp_path.exists() {
+        if let Ok(existing_content) = fs::read_to_string(&stamp_path) {
+            if existing_content == content {
+                // Content is identical, preserve mtime
+                return Ok(stamp_path);
+            }
+        }
+    }
+
+    // Write new content atomically
+    let temp_path = stamp_path.with_extension("tmp");
+    fs::write(&temp_path, content)?;
+    fs::rename(temp_path, &stamp_path)?;
+
+    Ok(stamp_path)
+}
+
+/// Generate all stamps for a detection run
+pub fn generate_detection_stamps(
```

Comment:
can this share more code with the cutout version? ideally adding a new model in the future does not need so much boilerplate